File: theory_of_encoding_compression_and_information_recovery/course_work/src/compressor/application/common/ports/transaction_manager.py
from abc import abstractmethod
from typing import Protocol


class TransactionManager(Protocol):
    """
    UoW-compatible interface for committing a business transaction.
    The implementation may be an ORM session, such as SQLAlchemy's.
    For more information, see: https://t.me/advice17/60
    """

    @abstractmethod
    async def commit(self) -> None:
        """
        Save all operations in database.
        :return: Nothing
        """
        ...

    @abstractmethod
    async def flush(self) -> None: ...

    @abstractmethod
    async def rollback(self) -> None:
        """
        Method that rolls back all operations.
        :return: Nothing
        """
        ...

File: сross_platform_software_systems/14/user_service/user_service/src/main/java/com/c3equalz/user_service/application/common/ports/TransactionManager.java
package com.c3equalz.user_service.application.common.ports;

import reactor.core.publisher.Mono;

/**
 * Unit of Work-compatible interface for committing a business transaction.
 * The implementation may be an ORM session, such as R2DBC's DatabaseClient
 * or Spring Data R2DBC's ReactiveTransactionManager.
 * <p>
 * For more information, see: https://t.me/advice17/60
 */
public interface TransactionManager {

    /**
     * Save all operations in database.
     * Commits the current transaction.
     *
     * @return Mono that completes when the commit is done
     */
    Mono<Void> commit();

    /**
     * Flush all pending operations to the database without committing.
     * This is useful for getting generated IDs or for validation purposes.
     *
     * @return Mono that completes when the flush is done
     */
    Mono<Void> flush();

    /**
     * Method that rolls back all operations.
     * Reverts all changes made in the current transaction.
     *
     * @return Mono that completes when the rollback is done
     */
    Mono<Void> rollback();
}

